lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,354 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <optional>
+
+#include "mlir/Bytecode/BytecodeReader.h"
+#include "mlir/Bytecode/BytecodeWriter.h"
+#include "mlir/Dialect/DLTI/DLTI.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/FunctionInterfaces.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Target/LLVMIR/Dialect/All.h"
+
+#include "revng/Pipeline/RegisterContainerFactory.h"
+
+#include "revng-c/mlir/Dialect/Clift/IR/Clift.h"
+#include "revng-c/mlir/Dialect/Clift/Utils/Helpers.h"
+#include "revng-c/mlir/Pipes/MLIRContainer.h"
+
+using namespace revng;
+using namespace revng::pipes;
+
+using mlir::FunctionOpInterface;
+using mlir::MLIRContext;
+using mlir::ModuleOp;
+using mlir::Operation;
+using mlir::SymbolOpInterface;
+using mlir::SymbolTable;
+
+using ContextPtr = std::unique_ptr<MLIRContext>;
+using OwningModuleRef = mlir::OwningOpRef<ModuleOp>;
+
+static mlir::Block &getModuleBlock(ModuleOp Module) {
+  revng_assert(Module);
+  revng_assert(Module->getNumRegions() == 1);
+  return Module->getRegion(0).front();
+}
+
+static decltype(auto) getModuleOperations(ModuleOp Module) {
+  return getModuleBlock(Module).getOperations();
+}
+
+static bool isTargetFunction(FunctionOpInterface F) {
+  return F->getAttrOfType<mlir::StringAttr>(FunctionEntryMDName) != nullptr;
+}
+
+template<typename R, typename C, typename P>
+static P visitHelper(R (C::*)(P));
+template<typename R, typename C, typename P>
+static P visitHelper(R (C::*)(P) const);
+
+// Helper for visiting module operations non-recursively.
+// Allows erasing the visited operation during visitation.
+template<typename Visitor>
+static void visit(ModuleOp Module, Visitor visitor) {
+  auto &OpList = getModuleOperations(Module);
+
+  auto Begin = OpList.begin();
+  const auto End = OpList.end();
+
+  while (Begin != End) {
+    using RequestedOpT = decltype(visitHelper(&Visitor::operator()));
```

Comment:
Why not a range-`for`?

Reply:
The iterator is advanced in the middle of the loop before calling the visitor. This is so that the visitor is able to safely erase the operation.

Reply:
Uhm, OK, just make sure that the underlying data structure you're iterating actually offers the kind of iterator invalidation guarantees you expect.

Reply:
I added explicit typing for `OpList` and an explanation of what's happening. See [here](https://github.com/revng/revng-c/pull/1/files#diff-24dded21127406a4fb4c2ef29c414dd57a94b946de77b22cf143ababf124cc4dR59).

Reply:
OK, I just want to make sure you double checked that.